File: go/94.go
/**
题目概述:
给定一个二叉树的根节点 root ，返回它的 中序 遍历。

示例 1：

输入：root = [1,null,2,3]
输出：[1,3,2]

进阶: 递归算法很简单，你可以通过迭代算法完成吗？


*/
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// 递归
func inorderTraversal(root *TreeNode) []int {
   var res []int
   var inner func(node *TreeNode)
   inner=func(node *TreeNode){
   	  if node ==nil{
		  return
	  }
   	  inner(node.Left)
   	  res=append(res,node.Val)
   	  inner(node.Right)
   }
   inner(root)
   return res
}

// 非递归（栈,迭代）
func inorderTraversal(root *TreeNode) []int {
     var res []int
     stack:=[]*TreeNode{}
     for root !=nil || len(stack)>0{
     	for root !=nil{
     		stack=append(stack,root)
     		root=root.Left
		}
     	top:=len(stack)-1
     	root=stack[top]
		stack=stack[:top]
     	res=append(res,root.Val)
     	root=root.Right
	 }
     return res
}

File: php/94.php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($val = 0, $left = null, $right = null) {
 *         $this->val = $val;
 *         $this->left = $left;
 *         $this->right = $right;
 *     }
 * }
 */
class Solution {

    /**
     * @param TreeNode $root
     * @return Integer[]
     */
    function inorderTraversal($root) {
      $res= [];
      $this->helper($root,$res);
      return $res;

    }

    // 递归
    function helper($root,&$res) {
      if($root==null) return;
      $this->helper($root->left,$res);
      $res[]=$root->val;
      $this->helper($root->right,$res);
    }

    // 迭代
     function inorderTraversal($root) {
          $res= [];
          $stack=[];
          while($root !=null || !empty($stack)){
             while($root !=null){
                $stack[]=$root;
                $root=$root->left;
             }

             $root=array_pop($stack);
             $res[]=$root->val;
             $root=$root->right;
          }
        return $res;
     }

}